### ISSUE TITLE
GC: Data may remain (up to 1GB) after garbage collecting everything

### ISSUE THREAD

**Issue by schomatis** [ipfs/go-ds-badger#54]
Regarding a future transition to Badger, its garbage collection process [bails out](https://github.com/dgraph-io/badger/blob/8b89769ed71e56cf7f80abdcfd3576efc97bd412/value.go#L891-L892) in the case there is only one value log file, and by default the size of a value log file is set to [1GB](https://github.com/ipfs/go-ipfs/blob/52301ce84121c1446de71b698219656189f12b2e/repo/fsrepo/datastores.go#L374-L375). This translates to the scenario where a user adds data and then "deletes" it (e.g., unpinning and garbage collecting it) and it will still have up to 1GB of disk taken with these values marked as deleted but not reclaimed by Badger's GC.

This is not an error, and for big players 1GB may be negligible, but it will raise a normal end user's attention (specially because now everything is deleted and the disk space gets freed), and should be taken in consideration (maybe in the form of clear documentation clarifying that, or catching the [`ErrNoRewrite`](https://github.com/ipfs/go-ds-badger/blob/bfda9737bd7f9463fd70b58b7bf8be418ab3cb4d/datastore.go#L265) error and printing a warning to the user when appropriate).

**Comment by BenLubar** [comment]
I ran `mkfifo badger.bak` and then `badger backup` and `badger restore` and I've got a LOT of data that's not being garbage collected:

```
ben@urist:/storage/ipfs/data-badgerdebug$ du -bsch badgerds badgerds-restore
81G     badgerds
22G     badgerds-restore
```

ipfs repo gc was run right before the btrfs snapshot was taken.

**Comment by magik6k** [comment]
Related badger issue - https://github.com/dgraph-io/badger/issues/767

**Comment by Stebalien** [comment]
Triage: STILL BROKEN! Even after the periodic GC and repeated GC fixes.

**Comment by jsign** [comment]
We had a similar experience (with badger v1 or v2). Forcing GC, or doing manual `badger flatten`, etc doesn't seem to reclaim space on log files.

The only thing that worked was doing a backup and restore (offline).

**Comment by Stebalien** [comment]
That is very annoying. We're now doing _repeated_ GC cycles until badger reports there is nothing to do and that's still not fixing it.

**Comment by jsign** [comment]
Sharing [this](https://github.com/dgraph-io/badger/issues/1228) topic which might be interesting to keep track of. My pain-points are related to may _vlogs_, but the solution might indirectly improve some part of the problem.

**Comment by Stebalien** [comment]
That's great to hear! However, it looks like we're going to need to switch to badger 2 to see those improvements. We're currently on badger 1 as we haven't tested badger 2 thoroughly.

**Comment by jsign** [comment]
@Stebalien, I had my own shot at having some easy reproducible program for this problem and created an issue on the badger repo. Details [here](https://github.com/dgraph-io/badger/issues/1297).

TL;DR: seems like `NumVersionsToKeep: 0`, `CompactL0OnClose: true`, `ValueLogFileSize: XXX` with a reasonable `GCDiscardRatio`, is bounding sizes in some direct proportion to `ValueLogFileSize`. So, knowing the right knots to affect that dimension is good to know. 

Using the program I created for the issue (`jsign/go-badger2-size`) these setups can be ran and filesize can be inspected. `master` for Badger v2, and `badgerv1` for Badger v1. In both versions, these knots seem to work.

That program is closing the database before taking file-size measurements. I've tried another version of measuring before closing, and still, the file size reduction is significant. Obviously, not as high as after closing due to L0 Compaction on close. 

Right tunning of these params might require some extra work to discover the perf cost and right tradeoff.

**Comment by RubenKelevra** [comment]
@jsign sounds like we could use those parameters as default and let the user tweak them if they want more performance (or tweak them in the server profiles).

I end up having a repo with 6 GB on a node with no data, which is quite confusing. I doubt that any of this data remaining will be helpful in the future to speedup fetching/storing different content.

Can you do a PR to implement those tweaks?

**Comment by jsign** [comment]
> @jsign sounds like we could use those parameters as default and let the user tweak them if they want more performance (or tweak them in the server profiles).
> 
> I end up having a repo with 6 GB on a node with no data, which is quite confusing. I doubt that any of this data remaining will be helpful in the future to speedup fetching/storing different content.
> 
> Can you do a PR to implement those tweaks?

Yes, that can be an option, but changing defaults might have some unexpected consequences for existing instances since they might spin-up again on new defaults. Being v0.* and its semantic on semver makes that tricky.

I'm already quite happy about finding a solution that _seems_ to be reproducible and _solves_ this problem. If we can have some extra confirmations that others can reproduce the solution I presented, maybe it's just giving it enough visibility in the readme or similar. 

Or changing the defaults, tagging a new version, and giving enough warning signs in the changelogs (which I doubt much people read).

**Comment by RubenKelevra** [comment]
> > @jsign sounds like we could use those parameters as default and let the user tweak them if they want more performance (or tweak them in the server profiles).
> > I end up having a repo with 6 GB on a node with no data, which is quite confusing. I doubt that any of this data remaining will be helpful in the future to speedup fetching/storing different content.
> > Can you do a PR to implement those tweaks?
> 
> Yes, that can be an option, but changing defaults might have some unexpected consequences for existing instances since they might spin-up again on new defaults. Being v0.* and its semantic on semver makes that tricky.
> 
> I'm already quite happy about finding a solution that _seems_ to be reproducible and _solves_ this problem. If we can have some extra confirmations that others can reproduce the solution I presented, maybe it's just giving it enough visibility in the readme or similar.
> 
> Or changing the defaults, tagging a new version, and giving enough warning signs in the changelogs (which I doubt much people read).

I agree that these changes need to be thoroughly tested. But since your PR wouldn't make it to 0.5.0 anyway, I don't see an issue changing the defaults.

We might want to remove the "loop until badger reports there's nothing to clean up, @Stebalien mentioned. But that's another story.

**Comment by RubenKelevra** [comment]
@Stebalien 

> That is very annoying. We're now doing _repeated_ GC cycles until badger reports there is nothing to do and that's still not fixing it.

Wondering if this is related to https://github.com/ipfs/go-ipfs/issues/7283 and https://github.com/ipfs/go-ipfs/issues/7273 - maybe we run the GC over and over again and BadgerDS doesn't return "there's nothing to do"?

Maybe we can print a warning if we run a GC event on Badger-DB on the console, to at least inform the user, what's going on.
 
> TL;DR: seems like `NumVersionsToKeep: 0`, `CompactL0OnClose: true`, `ValueLogFileSize: XXX` with a reasonable `GCDiscardRatio`, is bounding sizes in some direct proportion to `ValueLogFileSize`. So, knowing the right knots to affect that dimension is good to know.

Can we get those options exposed as experimental flags or something like this? To see what performance impact this has? :)

**Comment by RubenKelevra** [comment]
Another example (with more data):

```
[ipfs@heimdal-pacman-store ~]$ ipfs pin ls
[ipfs@heimdal-pacman-store ~]$ ipfs files ls
[ipfs@heimdal-pacman-store ~]$ time ipfs repo gc

real    0m16.464s
user    0m0.012s
sys     0m0.035s
[ipfs@heimdal-pacman-store ~]$ time ipfs repo stat --human
NumObjects: 4
RepoSize:   23 GB
StorageMax: 75 GB
RepoPath:   /home/ipfs/.ipfs
Version:    fs-repo@9

real    0m6.191s
user    0m0.000s
sys     0m0.147s
```

Performance - even on this small machine is great with badgerds, but there seems to be zero bytes cleanup after I deleted data with the current settings.

**Comment by CsterKuroi** [comment]
@RubenKelevra The same situation

**Comment by RubenKelevra** [comment]
> > @jsign sounds like we could use those parameters as default and let the user tweak them if they want more performance (or tweak them in the server profiles).
> > I end up having a repo with 6 GB on a node with no data, which is quite confusing. I doubt that any of this data remaining will be helpful in the future to speedup fetching/storing different content.
> > Can you do a PR to implement those tweaks?
> 
> Yes, that can be an option, but changing defaults might have some unexpected consequences for existing instances since they might spin-up again on new defaults. Being v0.* and its semantic on semver makes that tricky.

How about exposing those settings and make them accessible via the configuration file?

The configuration files won't be updated on existing setups, we could just edit the badgerds profile, to include the changes on new setups.

This way existing setups won't be affected at all.

> I'm already quite happy about finding a solution that _seems_ to be reproducible and _solves_ this problem. If we can have some extra confirmations that others can reproduce the solution I presented, maybe it's just giving it enough visibility in the readme or similar.

Can you look into creating the changes? Would really enjoy to test these out in my setup :)

**Comment by jarifibrahim** [comment]
Hey guys, we've closed https://github.com/dgraph-io/badger/issues/1228 with fixes in master. We'll do a new badger release soon. It would be very useful if you can test the master branch.

**Comment by Stebalien** [comment]
From what I can tell, that's mostly concerned with expiring keys, right?

**Comment by jarifibrahim** [comment]
No. It's for deleted and expired keys. There was a bug in compaction because of which we would keep deleted/expired keys. We no longer do that.

**Comment by jsign** [comment]
That sounds promising :)

**Comment by jarifibrahim** [comment]
We've had issues with deleted keys, move keys in badger for a long time (this issue was created 2 years ago). I'm glad we fixed it :)

Please do try it out. We're always looking for feedback :stars:

**Comment by jsign** [comment]
> We've had issues with deleted keys, move keys in badger for a long time (this issue was created 2 years ago). I'm glad we fixed it :)
> 
> Please do try it out. We're always looking for feedback 🌠

@jarifibrahim , do you have some release estimate? I `master` is available, but I'm not sure you have some special tests on releases to have more guarantees about safety.

**Comment by jarifibrahim** [comment]
>@jarifibrahim , do you have some release estimate? I know master is available, but I'm not sure you have some special tests on releases to have more guarantees about safety.

We wanted to do Badger release last week but some other things came up (we're a small team of engineers working on badger and dgraph). I'll try to get a new patch version for v1.x and v2.x before end of this week (worst case end of next week).

Most of the tests for badger run as part of the CI. We have a bank test https://github.com/dgraph-io/badger/blob/master/badger/cmd/bank.go which checks for transactional issues and runs for (4h+4h) 8 hours at midnight every day.

Which version of badger are you using?

**Comment by Stebalien** [comment]
go-ipfs is using v1.6.1 and has been slowly adding support for v2.

**Comment by CsterKuroi** [comment]
Any update ?

**Comment by Stebalien** [comment]
Not yet. The next step here is to copy the badger plugin (https://github.com/ipfs/go-ipfs/tree/master/plugin/plugins/badgerds) as "badger2ds", and make it use github.com/ipfs/go-ds-badger2. Want to submit a PR to go-ipfs?

**Comment by CsterKuroi** [comment]
https://github.com/dgraph-io/badger/releases/tag/v1.6.2

**Comment by RubenKelevra** [comment]
Soooo ... how is this looking? Is it fixed? Can it be closed?

**Comment by Stebalien** [comment]
Not in badger v1. This has been fixed in v3 (and maybe badger v2 but that has other issues).

**Comment by RubenKelevra** [comment]
@Stebalien but this PR sounds like this bug should be fixed: https://github.com/dgraph-io/badger/pull/1354

**Comment by Stebalien** [comment]
That was a v2 fix, not a v1 fix.

**Comment by dokterbob** [comment]
:'(

Such a pity - we saw huge performance gains on ipfs-search but it is eating up storage an alarming rate. Will have to switch back to flatfs, for now. 
```
$ ipfs repo stat --human && ipfs repo gc -q | wc -l && ipfs repo stat --human
NumObjects: 1102
RepoSize:   904 GB
StorageMax: 10 GB
RepoPath:   /var/lib/ipfs
Version:    fs-repo@11
1143
NumObjects: 111
RepoSize:   904 GB
StorageMax: 10 GB
RepoPath:   /var/lib/ipfs
Version:    fs-repo@11
```

**Comment by dokterbob** [comment]
Note: after shutting down ipfs and calling `badger flatten`, garbage collection does seem to work.

```
$ go get github.com/dgraph-io/badger/...@v1.6.2
$ badger flatten --dir /var/lib/ipfs/badgerds
$ ipfs repo stat --human && ipfs repo gc -q | wc -l && ipfs repo stat --human
NumObjects: 1127
RepoSize:   320 GB
StorageMax: 10 GB
RepoPath:   /var/lib/ipfs
Version:    fs-repo@11

1118
NumObjects: 2462
RepoSize:   89 GB
StorageMax: 10 GB
RepoPath:   /var/lib/ipfs
Version:    fs-repo@11
$ ipfs repo stat --human && ipfs repo gc -q | wc -l && ipfs repo stat --human
NumObjects: 4280
RepoSize:   961 MB
StorageMax: 10 GB
RepoPath:   /var/lib/ipfs
Version:    fs-repo@11
4271
NumObjects: 10
RepoSize:   961 MB
StorageMax: 10 GB
RepoPath:   /var/lib/ipfs
Version:    fs-repo@11
```
